File: elixir/freelancer-rates/lib/freelancer_rates.ex
defmodule FreelancerRates do
  @daily_rate 8.0
  @monthly_billable_days 22

  def daily_rate(hourly_rate) do
    hourly_rate * @daily_rate
  end

  def apply_discount(before_discount, discount) do
    before_discount * (100 - discount) / 100
  end

  def daily_discounted(hourly_rate, discount) do
    hourly_rate
    |> daily_rate
    |> apply_discount(discount)
  end

  def monthly_rate(hourly_rate, discount) do
    (daily_discounted(hourly_rate, discount) * @monthly_billable_days)
    |> ceil
  end

  def days_in_budget(budget, hourly_rate, discount) do
    (budget / daily_discounted(hourly_rate, discount))
    |> Float.floor(1)
  end
end

File: cpp/freelancer-rates/freelancer_rates.cpp
#include <cmath>

double daily_rate(double hourly_rate) { 
    return hourly_rate * 8.0;
}

double apply_discount(double before_discount, double discount) {
    return before_discount * (1.0 - discount * 0.01);
}

int monthly_rate(double hourly_rate, double discount) {
    return ceil(apply_discount(daily_rate(hourly_rate) * 22.0, discount));
}

int days_in_budget(int budget, double hourly_rate, double discount) {
    return floor(budget / apply_discount(daily_rate(hourly_rate), discount));
}
